### DISCUSSION TITLE
More explanation of autoUnref option in 4.0 client

### DISCUSSION THREAD

**Discussion by qiulang** [socketio/socket.io#4288] [Q&A] [answered]
As explained here https://socket.io/docs/v3/migrating-from-3-x-to-4-0/#autounref-option,

With `autoUnref`set to true (default: false), the Socket.IO client will allow the program to exit if there is no other active timer/socket in the event system:

```
const socket = io({
  autoUnref: true
});
```

But even in client 2.x if I close all the active sockets the program will exit, so what does these words "the Socket.IO client will allow the program to exit if there is no other active timer/socket in the event system" mean exactly ?

**Comment by darrachequesne** [comment, marked as answer]
Hi! Setting `autoUnref` to `true` will allow the program to exit, even if the socket is currently connected.

```js
const socket = io({
  autoUnref: true
});

setTimeout(() => {
  // program will now exit
}, 1000);

```

Without it, you will indeed need to call `socket.disconnect()`.

**Reply by qiulang** [reply]
Thanks for the reply. But I was confused with the use case of this. When there is a socket connected why would I want the program to exit? 

One use case I can think of maybe is the graceful shutdown. But even in that case I can still close the socket to let program exit. So what is the use case ?

**Reply by darrachequesne** [reply]
@qiulang please see the discussion there: https://github.com/socketio/engine.io-client/pull/652#issuecomment-784405718

**Reply by qiulang** [reply]
Thanks for the reply. I have to say I still agree with your there "I'm not sure about merging this. I think I'd rather have the users explicitly call close(), instead of implicitly closing the connection when there is nothing left in the event loop." 😅

**Reply by qiulang** [reply]
The argument that "but the lack of an unref() for use by the library developer forces the API consumer to handle socket closing" is not convincing to me. I will argue that if you (the consumer) use socket.io, you should be aware that you need to close it when you are done just like when you open a database connection you should close the connection when you are finished.

**Reply by qiulang** [reply]
@darrachequesne Now I understand this feature more, so for a short-lived script it is convenient. But I think I need to point out that ioredia author seemed to reject the idea at the end as I commented on https://github.com/socketio/engine.io-client/pull/652
